### PR TITLE
Axon and Storm wget improvements (SYN-4675)

### PR DESCRIPTION
**Pull request by vEpiphyte**
- Axon.wget() now includes the request and redirect history information
- $lib.axon.wget now includes the original URL in its response
- Storm wget / $lib.axon.urlfile now makes inet:urlfile nodes based on the original request URL to preserve pivoting 1:1 from a URL to the urlfile node.
- Storm wget / $lib.axon.urlfile now makes inet:urlredir nodes based on the redirect history information. The first node in the chain uses the original request url to preserve pivoting 1:1 from the url to the redirection chain.

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -9,10 +9,21 @@ NEXTVERS - YYYY-MM-DD
...
+  about the original request and URL redirect history in their response data.
+  The Storm ``wget`` command now creates ``inet:urlredir`` nodes based on the
+  redirect history when that is present.
+  (`#3011 <https://github.com/vertexproject/synapse/pull/3011>`_)
 
 Bugfixes
 --------
+- The Storm ``wget`` command created ``inet:urlfile`` nodes with the ``url``
+  property of the resolved URL from ``aiohttp``. This broke made it so that
+  a user could not pivot from a ``inet:url`` node which had a URL encoded
```
Comment (Cisphyx) [inline CHANGELOG.rst:22]: "an" inet:url

File: CHANGELOG.rst [hunk 2]
```diff
@@ -9,10 +9,21 @@ NEXTVERS - YYYY-MM-DD
...
+  parameter strin to the resulting ``inet:urlfile`` node. The
```
Comment (Cisphyx) [inline CHANGELOG.rst:23]: "strin"

File: synapse/axon.py [hunk 3]
```diff
@@ -1574,6 +1580,47 @@ async def wput(self, sha256, url, params=None, headers=None, method='PUT', ssl=T
...
+    async def _flatten_clientresponse(self,
+                                      resp: aiohttp.ClientResponse,
+                                      toplevel=True,
+                                      ) -> dict:
+        info = {
+            'ok': True,
+            'url': str(resp.url),
+            'code': resp.status,
+            'headers': dict(resp.headers),
+            'request': {
```
Comment (Cisphyx) [inline synapse/axon.py:1592]: can probably skip setting this here since we set it again below

File: synapse/axon.py [hunk 4]
```diff
@@ -1574,6 +1580,47 @@ async def wput(self, sha256, url, params=None, headers=None, method='PUT', ssl=T
...
+                'url': str(resp.request_info.url),
+                'headers': dict(resp.request_info.headers),
+                'method': str(resp.request_info.method),
+            }
+        }
+
+        hashset = s_hashset.HashSet()
```
Comment (Cisphyx) [inline synapse/axon.py:1599]: could move this into the toplevel block since it only gets used there

File: synapse/axon.py [hunk 5]
```diff
@@ -1607,7 +1654,13 @@ async def wget(self, url, params=None, headers=None, json=None, body=None, metho
                         'sha1': <str> - The SHA1 hash of the response body.
                         'sha256': <str> - The SHA256 hash of the response body.
                         'sha512': <str> - The SHA512 hash of the response body.
+                    },
+                    'request': {
+                        'url': The request URL. This is a url-decoded string.
```
Comment (Cisphyx) [inline synapse/axon.py:1659]: also a `real_url` in the dict

File: synapse/axon.py [hunk 6]
```diff
@@ -621,7 +621,13 @@ async def wget(self, url, params=None, headers=None, json=None, body=None, metho
...
```
Comment (Cisphyx) [inline synapse/axon.py:626]: also has a `real_url`

File: synapse/lib/stormtypes.py [hunk 7]
```diff
@@ -2062,27 +2064,52 @@ async def urlfile(self, *args, **kwargs):
...
+
+            # We skip the first entry in history, since that URL is the original URL
+            # having been redirected. The second+ history item represents the
+            # requested URL. We then capture the last part of the chain in our list.
+            # The recorded URLs after the original_url are all the resolved URLS,
+            # since Location headers may be partial paths and this avoids needing to
+            # do url introspection that has already been done by the Axon.
+
+            for info in history[1:]:
+                url = info.get('url')
```
Comment (Cisphyx) [inline synapse/lib/stormtypes.py:2104]: should this use `real_url`?

File: synapse/axon.py (hunk 5 above)
Comment (vEpiphyte) [reply synapse/axon.py:1659]: I need to remove that thanks

File: synapse/axon.py (hunk 4 above)
Comment (vEpiphyte) [reply synapse/axon.py:1599]: moved

File: CHANGELOG.rst [hunk 8]
```diff
@@ -12,9 +12,26 @@ Features and Enhancements
 - Add ``synapse.utils.stormcov``, a Coverage.py plugin for measuring code
   coverage of Storm files.
   (`#2961 <https://github.com/vertexproject/synapse/pull/2961>`_)
-
-Bugfixes
---------
+- The ``Axon.wget()`` API node includes HTTP request history, which is added
```
Comment (Cisphyx) [inline CHANGELOG.rst:15]: 'API node'

File: CHANGELOG.rst [hunk 9]
```diff
@@ -9,20 +9,25 @@ NEXTVERS - YYYY-MM-DD
...
 -------------------------
-- The ``Axon.wget()`` and ``$lib.axon.wget()`` APIs now include information
-  about the original request and URL redirect history in their response data.
-  The Storm ``wget`` command now creates ``inet:urlredir`` nodes based on the
-  redirect history when that is present.
+- The ``Axon.wget()`` API node includes HTTP request history, which is added
+  when the API encounters redirects. The ``$lib.axon.wget()`` Storm API now
+  includes information about the original request URL. This data is now used
+  to create ``inet:urlredir`` nodes, such as when the Storm ``wget`` command
+  is used to retreive a file.
```
Comment (Cisphyx) [inline CHANGELOG.rst:19]: retrieve

File: CHANGELOG.rst [hunk 10]
```diff
@@ -9,20 +9,25 @@ NEXTVERS - YYYY-MM-DD
...
 Bugfixes
 --------
 - The Storm ``wget`` command created ``inet:urlfile`` nodes with the ``url``
-  property of the resolved URL from ``aiohttp``. This broke made it so that
-  a user could not pivot from a ``inet:url`` node which had a URL encoded
-  parameter strin to the resulting ``inet:urlfile`` node. The
-  ``inet:urlfile`` nodes are now made with the original request URL to allow
-  that pivoting to occur.
+  property of the resolved URL from ``aiohttp``. This made it so that a user
+  could not pivot from a ``inet:url`` node which had an URL encoded  parameter
```
Comment (Cisphyx) [inline CHANGELOG.rst:26]: 'an' inet:url, extra space between 'encoded  parameter'

Comment (Cisphyx) [reply CHANGELOG.rst:26]: 'a' URL

File: CHANGELOG.rst [hunk 11]
```diff
@@ -9,20 +9,25 @@ NEXTVERS - YYYY-MM-DD
...
+  string to the resulting ``inet:urlfile`` node. The ``inet:urlfile`` nodes
+  are now made with the original request URL to allow that pivoting to occur.
+  (`#3011 <https://github.com/vertexproject/synapse/pull/3011>`_)
+- The ``Axon.wget()`` and ``$lib.axon.wget()`` APIs returned URLs in the
+  ``url`` field of their responses which did contain fragment identifiers.
```
Comment (Cisphyx) [inline CHANGELOG.rst:31]: missing 'not'

**Review by Cisphyx — APPROVED** [review]

**Review by Cisphyx — APPROVED** [review]

**Pull request merged**